graph/src/blockchain/firehose_block_stream.rs

```diff
@@ -0,0 +1,252 @@
+use futures03::{FutureExt, Stream, StreamExt};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use crate::prelude::*;
+use crate::sf::endpoints::FirehoseEndpoint;
+
+use super::block_stream::{BlockStream, BlockStreamEvent, FirehoseMapper};
+use super::Blockchain;
+use crate::sf::bstream;
+
+pub struct FirehoseBlockStreamContext<C>
+where
+    C: Blockchain,
+{
+    cursor: Option<String>,
+    mapper: Arc<C::FirehoseMapper>,
+    node_id: NodeId,
+    subgraph_id: DeploymentHash,
+    filter: Arc<C::TriggerFilter>,
+    start_blocks: Vec<BlockNumber>,
+    logger: Logger,
+}
+
+impl<C: Blockchain> Clone for FirehoseBlockStreamContext<C> {
+    fn clone(&self) -> Self {
+        Self {
+            cursor: self.cursor.clone(),
+            mapper: self.mapper.clone(),
+            node_id: self.node_id.clone(),
+            subgraph_id: self.subgraph_id.clone(),
+            filter: self.filter.clone(),
+            start_blocks: self.start_blocks.clone(),
+            logger: self.logger.clone(),
+        }
+    }
+}
+
+enum BlockStreamState {
+    Disconneted,
+    Connecting(
+        Pin<
+            Box<
+                dyn futures03::Future<
+                    Output = Result<tonic::Streaming<bstream::BlockResponseV2>, anyhow::Error>,
+                >,
+            >,
+        >,
+    ),
+    Connected(tonic::Streaming<bstream::BlockResponseV2>),
+}
+
+pub struct FirehoseBlockStream<C: Blockchain> {
+    endpoint: Arc<FirehoseEndpoint>,
+    state: BlockStreamState,
+    ctx: FirehoseBlockStreamContext<C>,
+    connection_attempts: u64,
+}
+
+impl<C> FirehoseBlockStream<C>
+where
+    C: Blockchain,
+{
+    pub fn new(
+        endpoint: Arc<FirehoseEndpoint>,
+        cursor: Option<String>,
+        mapper: Arc<C::FirehoseMapper>,
+        node_id: NodeId,
+        subgraph_id: DeploymentHash,
+        filter: Arc<C::TriggerFilter>,
+        start_blocks: Vec<BlockNumber>,
+        logger: Logger,
+    ) -> Self {
+        FirehoseBlockStream {
+            endpoint,
+            state: BlockStreamState::Disconneted,
+            ctx: FirehoseBlockStreamContext {
+                cursor,
+                mapper,
+                node_id,
+                subgraph_id,
+                logger,
+                filter,
+                start_blocks,
+            },
+            connection_attempts: 0,
+        }
+    }
+}
+
+impl<C: Blockchain> BlockStream<C> for FirehoseBlockStream<C> {}
+
+impl<C: Blockchain> Stream for FirehoseBlockStream<C> {
+    type Item = Result<BlockStreamEvent<C>, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
I find that reasoning about `poll` implementations is harder than reasoning about async/await and stream combinators, and the manual `Stream` implementation is one thing I dislike about the current block stream. I see you already hit some complications here such as in `schedule_error_retry`.

I'd suggest trying to write this stream without a manual implementation. [try_unfold](https://docs.rs/futures/0.3.17/futures/stream/fn.try_unfold.html) might be a good way to implement a state machine with async/await instead of messing with `Poll`.

Reply:
This can be left as a to-do for future PRs.

Reply:
I totally agree, we found it really hard also and were trying to look at alternative. The `try_unfold` seems pretty good, I'll take a look and revisit later.

Reply:
How one can use the combinator like `try_unfold` and "implement" the `Stream<Item = ...>` trait?

Like I have my struct that right now implement the `poll` method, what is the recipe so that I still implement the trait but use `try_unfold` as the implementation?

```
impl<C: Blockchain, F: FirehoseMapper<C>> Stream for FirehoseBlockStream<C, F> {
   ...?
}
```

Is the trick to actual create a stream "object" using `try_unfold` and then use that to return the `poll` value?